src/metabase/api/dashboard.clj

```diff
@@ -215,14 +216,97 @@
+    (reduce (fn [m [id card]]
+              (assoc m id
+                     (if (:dataset card)
+                       card
+                       (api.card/create-card!
+                        (cond-> (assoc card :collection_id dest-coll-id)
+                          same-collection?
+                          (update :name #(str % " -- " (tru "Duplicate"))))))))
+            {:uncopied discard}
+            copy)))
+
+(defn update-cards-for-copy
+  "Update ordered-cards in a dashboard for copying. If shallow copy, returns the cards. If deep copy, replaces ids with
+  id from the newly-copied cards. If there is no new id, it means user lacked curate permissions for the cards
+  collections and it is omitted. Dashboard-id is only needed for useful errors."
+  [dashboard-id ordered-cards deep? id->new-card]
+  (when (and deep? (nil? id->new-card))
+    (throw (ex-info (tru "No copied card information found")
+                    {:user-id api/*current-user-id*
+                     :dashboard-id dashboard-id})))
+  (if-not deep?
+    ordered-cards
+    (keep (fn [dashboard-card]
+            (cond
+              ;; text cards need no manipulation
+              (nil? (:card_id dashboard-card))
+              dashboard-card
+
+              ;; if we didn't duplicate, it doesn't go in the dashboard
+              (not (id->new-card (:card_id dashboard-card)))
+              nil
+
+              :else
+              (let [new-id (fn [id]
+                             (-> id id->new-card :id))]
+                (-> dashboard-card
+                    (update :card_id new-id)
+                    (assoc :card (-> dashboard-card :card_id id->new-card))
+                    (m/update-existing :parameter_mappings
+                                       (fn [pms]
+                                         (map (fn [pm]
+                                                (m/update-existing pm :card_id new-id))
+                                              pms)))
+                    (m/update-existing :series
+                                       (fn [series]
+                                         (keep (fn [card]
+                                                 (when-let [id' (new-id (:id card))]
+                                                   (assoc card :id id')))
+                                               series)))))))
+          ordered-cards)))
 
 (api/defendpoint POST "/:from-dashboard-id/copy"
   "Copy a Dashboard."
-  [from-dashboard-id :as {{:keys [name description collection_id collection_position], :as _dashboard} :body}]
-  {name                (s/maybe su/NonBlankString)
-   description         (s/maybe s/Str)
-   collection_id       (s/maybe su/IntGreaterThanZero)
-   collection_position (s/maybe su/IntGreaterThanZero)}
+  [from-dashboard-id :as {{:keys [name description collection_id collection_position
+                                  is_deep_copy], :as _dashboard} :body}]
```

Comment:
```suggestion
                                  is_deep_copy], :as _dashboard, :or {is_deep_copy false}} :body}]
```

Might be a good idea to have a default value here. It's better for readability IMO.

Reply:
i find nil destructuring just as readable. no strong opinion but this feels redundant to me. Of course the default value is nil.